### ISSUE TITLE
Keychords behave incorrectly

### ISSUE THREAD

**Issue by alethiophile** [qtile/qtile#3196]
Currently, the behavior of the KeyChord mapping style is to, on receiving the keychord entry keystroke, unmap all other keys except those which are mapped in the keychord; then remain in this keychord mode until a key that is mapped in the keychord is pressed (or longer, if the keychord sets a mode), passing all unmapped keys through to the application. This behavior is inconsistent with all other software that uses keychord mappings, and is basically incorrect.

The correct behavior for a keychord (that does not set a mode) is to, on receiving the keychord opening keystroke, _unconditionally_ grab the next key; then, if that key is not mapped in the keychord, either display an error or do nothing, but in any case, revert immediately back to the root keymap. This is the behavior of all other software I can think of that maps keychords, including StumpWM, Emacs and screen/tmux. It's also fairly clearly the correct behavior: no one ever actually wants to first open a keychord, then press other keys that go through to the application, then press a keychord-mapped key to complete the keychord action. Functionally speaking, keychords are always pressed in direct succession, with no other keys between. Thus, opening a keychord and then pressing a key that is unmapped in that keychord can be presumed to be an error, and so should be called out to the user and should not leave the system in an unexpected mode (here, with the keychord still open, thus ignoring keys in the root keymap). (This all applies to keychords that do not set a mode; modal activation is rarer and has fewer clear precedents, so I don't know exactly what the behavior should be there.)

I intend to submit a pull request regarding this issue sometime in the next week or so.

**Comment by m-col** [comment]
This seems reasonable, but the design of the keychords seems as though this was intential. I am 90% sure this specific behaviour has been discussed before, however unfortunately searching on github isn't helping me find the discussion. I'd like to hear the thoughts of others who were more involved with the keychord code, @ep12, perhaps @tych0, do you have thoughts on this?

**Comment by alethiophile** [comment]
I see some prior discussion along these lines in https://github.com/qtile/qtile/pull/2121. Someone there had the objection that the behavior shouldn't be changed on compatibility grounds.

As far as I'm concerned, this is just a bug, and preserving a bug just because that's the existing behavior is not sound.

**Comment by ep12** [comment]
I think you're talking about my comment there. That wasn't really a compatibility issue but rather my personal use case that would make use of the current behaviour (and the reason for me preferring an option). Actually, even with the suggested behaviour I could still use the default bindings of xournalpp by adding (a lot of) dummy bindings to the mode that just send the stroke through to the application.

Nevertheless I'm fine with that behaviour to change as other programs behave as you described it and I'm happy to see you've already started a pr.

**Comment by elParaguayo** [comment]
> I think you're talking about my comment there. That wasn't really a compatibility issue but rather my personal use case that would make use of the current behaviour (and the reason for me preferring an option). Actually, even with the suggested behaviour I could still use the default bindings of xournalpp by adding (a lot of) dummy bindings to the mode that just send the stroke through to the application.
> 
> Nevertheless I'm fine with that behaviour to change as other programs behave as you described it and I'm happy to see you've already started a pr.

Just an observation but the fact that you would need to add a load of dummy binding to retain your desired behaviour does show that some people may like the current implementation.

I'm largely indifferent to this but I would just recommend some acknowledgement that one person's "bug" may be another's "feature". In scenarios like that, I'm a bit more wary about completely changing behaviour without users having the option to retain the current behaviour). Those are the scenarios where we get a raft of new issues posted about something being broken.

@alethiophile - all that being said, I think your explanation of the behaviour makes total sense and should probably be the default behaviour (and thank you for submitting the PR). However, it may be helpful to have an option to retain the current behaviour so that other users can make a quick change rather than creating a load of dummy bindings to replicate that behaviour.

**Comment by alethiophile** [comment]
Regarding "normal" (non-modal) keychords, I think there isn't much of a possible use case for retaining the existing behavior, because rather than pressing <keychord key> <keys that go through to application> <key bound in keychord>, you can always just move the inner passed-through keys either before or after the keychord, and have the same effect. I think having a configuration option for this is probably not worth the extra code complexity.

For modal keychords, the potential use case is a lot stronger -- being able to mix application-passed keys and mode-bound keys freely can definitely be useful. (Though one annoyance with doing this is that Escape is hard-coded to end the keychord, and so can't ever go to the application during this. I'm not sure if this is worth changing.) Being able to set this option on modal keychords might be a good idea; I'll see if this is workable in the PR.

**Comment by elParaguayo** [comment]
That makes sense. Thanks.

**Comment by OlMon** [comment]
@alethiophile Hej Hej,

Just wanted to ask, if this is still coming? From the discussion it seems that this is a behavior that will be implemented. I see a PR that does not seems to be merged.

Can I get an update on this? 
Thanks a lot!

Cheers,
OlMon

**Comment by OlMon** [comment]
#3235 Seems to partially implement a "grab_keyboard" parameter, but there were some Wayland problems as far as I understood. The PR went stale and was closed by the bot just as this issue.

**Comment by alethiophile** [comment]
When I left this, the functionality was complete and as far as I can tell it worked both on X and Wayland. I was trying to write new tests for this functionality, but the X test harness was being uncooperative and I ran out of time to work on it.

In the interim, I stopped using qtile day-to-day, so there was less impetus to come back to it. I may yet come back and do more work on this, but the timing is unpredictable.

As it stands, my keychords branch from that PR should have a working implementation (but I have no idea if it will merge cleanly onto the current head).
